src/odoc/bin/main.ml

```diff
@@ -601,81 +601,49 @@ end = struct
-               ~f:(fun (pkgname, _) -> pkgname = curpkgnane)
-               page_roots)
+    match (current_package, detected_package) with
+    | Some curpkgnane, Some detected_package when detected_package <> curpkgnane
+      ->
+        Error
+          (`Msg
+            "The package name specified with --current-package is not \
+             consistent with the packages passed as a -P")
+    | _, (Some _ as r) (* we have equality or only detected package *) -> Ok r
+    | (Some given as g), None ->
+        if not (List.exists ~f:(fun (pkgname, _) -> pkgname = given) page_roots)
         then
           Error
             (`Msg
               "The package name specified with --current-package do not match \
                any package passed as a -P")
-        else
-          match detected_package with
-          | Some dpkg when dpkg <> curpkgnane ->
-              Error
-                (`Msg
-                  "The package name specified with --current-package is not \
-                   consistent with the packages passed as a -P")
-          | _ -> Ok current_package)
-    | None -> Ok detected_package
-
-  let current_package_of_page ~current_package page_roots input =
-    match find_root_of_input page_roots input with
-    | Ok detected_package ->
-        validate_current_package ?detected_package page_roots current_package
-    | Error `Not_found ->
-        Error (`Msg "The input file must be part of a directory passed as -P")
-
-  let is_page input =
-    input |> Fpath.filename |> Astring.String.is_prefix ~affix:"page-"
-
-  let is_asset input =
-    input |> Fpath.filename |> Astring.String.is_prefix ~affix:"asset-"
+        else Ok g
+    | None, None -> Ok None
+
+  let find_current_package ~current_package page_roots input =
+    let detected_package = find_root_of_input page_roots input in
+    validate_current_package ?detected_package page_roots current_package
 
   let link directories page_roots lib_roots input_file output_file
       current_package warnings_options open_modules =
     let input = Fs.File.of_string input_file in
     let output = get_output_file ~output_file ~input in
-    (if
-       not
-         (Antichain.check
-            (List.rev_append lib_roots page_roots |> List.map ~f:snd))
-     then
-       Error
-         (`Msg "Arguments given to -P and -L cannot be included in each others")
-     else Ok ())
-    >>= fun () ->
-    let is_page_or_asset = is_page input || is_asset input in
-    (if is_page_or_asset then Ok None
-     else current_library_of_input lib_roots input)
-    >>= fun current_lib ->
-    (if is_page_or_asset then
-       current_package_of_page ~current_package page_roots input
-     else validate_current_package page_roots current_package)
+    Antichain.check (page_roots |> List.map ~f:snd) ~opt:"-P" >>= fun () ->
+    Antichain.check (lib_roots |> List.map ~f:snd) ~opt:"-L" >>= fun () ->
+    let current_lib = current_library_of_input lib_roots input in
```

Comment:
We might as well just use ` find_root_of_input` (clearer imho) if `current_library_of_input` is just an alias